### ISSUE TITLE
Analyse FFF file from old Simplify3D

### ISSUE THREAD

**Issue by HadiSDev** [KeltE/Flashforge_Dreamer_Cura-definition_concept#1]
I found this .fff profile from Simplify3D. We can use it to create the cura profile

[FlashForge Dreamer.zip](https://github.com/KeltE/Flashforge_Dreamer_Cura-definition_concept/files/6184149/FlashForge.Dreamer.zip)

**Comment by KeltE** [comment]
Yere can see the offset and that T0 is probably on the right. The only thing I don't understand is why it is necessary to set a separate offset when reading the start in G-code EEPROM from memory to offset. And flashprint also does not generate M612 line. **(M612 X34.5 Y0.0 ; Set toolhead offsets)** Then it's a mystery why Cura adds the T0 command before starting the G-code for two extruded printers. And as I read then the Printer refuses to continue.
 
Then I still have to test if the printer works without the **M118** command.

I found a post on such a forum:
 
 **As you stated in a previous comment, command M118 is a proprietary thingy from FF with the following syntax:
M118 {X Y Z} (max size of the printed area with object and e.g. a wall) {T0} {T1} {D1/D2}
XYZ coordinates can be omitted afaik, T0 and T1 are left and right extruder respectively, so for a single-left print you'd only state T1 - and this is where the interesting part is - D1 (mirror) and D2 (duplication) are actually the key parameters to enable duplication/mirror mode.
**In these two modes, the print bed origin shifts to 66.5x125 (resulting in a print bed size of 133x250 mm). The bed is mirrored (happens through the printer firmware if D1 or D2 are asserted) with a keepout zone of x = 34 mm between those two mirror patches (so the two printheads won't collide).
I talked with FlashForge about this and they confirmed this.****
[https://www.thingiverse.com/thing:4574596/comments](url)

**Flashprint gives me this for duplicate mode on a simple box:
M118 X16.60 Y16.60 Z11.55 T0 T1 D2**

**Flashprint gives me this for mirror mode on a simple box:
M118 X16.60 Y16.60 Z11.55 T0 T1 D1**

**Comment by HadiSDev** [comment]
Yea okay it is a bit more tricky: 
* I think I found a cura solution for this problem

**Comment by KeltE** [comment]
I forgot to mention. Since the Printer is waiting for the command "**M108 - Select Active Extruder**" then I have attached the start g-code to the extruder file. And extruder 1 file has **X offset -34mm**.

**Comment by HadiSDev** [comment]
I think the easiest way to get startet is to take an example stl make flashprint create the gcode and then compare it with our profile

**Comment by KeltE** [comment]
I did so in comparison. The tool change seemed to work. But the offset section is still open. Since the dreamer has both tools on the same carrier, you should only add the correct offset to the correct tool.
And I'll try tomorrow when I remove the m118 line from the g-code. What then. Because I use Prusaslicer and it doesn't change the M118 line at all and the printer does well. The M118 seems to be the most important for IDEX printers.

**Comment by HadiSDev** [comment]
I am waiting for a small part to arrive for my Dreamer. After that I can begin to test it with a small print

**Comment by KeltE** [comment]
Hopefully the profile will work. Then it will be easier in the future.
I also have an update up front for my printer. Since there is now a modified firmware, I bought a microswiss all metal hotend.

**Comment by KeltE** [comment]
> I think the easiest way to get startet is to take an example stl make flashprint create the gcode and then compare it with our profile

HadiSDev have you researched and tested this profile?
https://github.com/eugr/Flashforge-for-Cura/issues

**Comment by HadiSDev** [comment]
Not yet. Still waiting for my Ziflex bed. Also sorry if I don't respond quickly. I am studying at the same time 🤣

**Comment by HadiSDev** [comment]
The repo you added is for the creator pro. We can still use it. Just need to change the parameters

**Comment by KeltE** [comment]
There is nowhere to rush with it. The most important things before. I also don't have much time with this research after work. At the weekend I generally spend more time. I just had an idea at work and wrote so that I wouldn't forget later. Can use the information there.

**Comment by HadiSDev** [comment]
Still waiting on the bed. Do you have a .stl to test on? Or should I just try the boat one?

**Comment by KeltE** [comment]
I've done them or made calibration cubes. [https://www.thingiverse.com/thing:1545913](url)
Boats are enough for the world. But first you just have to test that the printer is going in the right place and both heads are heating up and cooling down properly. For some reason the previous profiles I've researched have a problem that both heads heat up when printing with one head. And how it behaves when printing with two heads. Will switch from one to another and the distances would be correct, etc.

**Comment by HadiSDev** [comment]
Okay I got the flexible bed installed now. Which profile did you want me to test again? :)

**Comment by KeltE** [comment]
This profile: Flashforge_Dreamer_Cura-definition_concept. But for this to work it must also have the Flashforge_Dreamer_NX_definition copied.

**Comment by HadiSDev** [comment]
2021-04-06 23:14:09,638 - ERROR - [MainThread] UM.Settings.ContainerStack.deserialize [429]: When trying to deserialize Flashforge Dreamer, we received an unknown container ID (dreamer)
2021-04-06 23:14:09,641 - ERROR - [MainThread] UM.Logger.logException [106]: Exception: Error in the format of container Flashforge Dreamer: When trying to deserialize Flashforge Dreamer, we received an unknown container ID (dreamer)
2021-04-06 23:14:09,645 - ERROR - [MainThread] UM.Logger.logException [110]: Traceback (most recent call last):
2021-04-06 23:14:09,650 - ERROR - [MainThread] UM.Logger.logException [110]:   File "X:\4.8-exe\build\inst\lib\python3.5\site-packages\UM\Settings\ContainerRegistry.py", line 203, in findContainers
2021-04-06 23:14:09,655 - ERROR - [MainThread] UM.Logger.logException [110]:   File "C:\Program Files\Ultimaker Cura 4.8.0\plugins\LocalContainerProvider\LocalContainerProvider.py", line 75, in loadContainer
2021-04-06 23:14:09,659 - ERROR - [MainThread] UM.Logger.logException [110]:     container.deserialize(f.read(), file_path)
2021-04-06 23:14:09,662 - ERROR - [MainThread] UM.Logger.logException [110]:   File "X:\4.8-exe\build\inst\lib\python3.5\site-packages\cura\Settings\CuraContainerStack.py", line 309, in deserialize
2021-04-06 23:14:09,665 - ERROR - [MainThread] UM.Logger.logException [110]:   File "X:\4.8-exe\build\inst\lib\python3.5\site-packages\UM\Settings\ContainerStack.py", line 430, in deserialize
2021-04-06 23:14:09,668 - ERROR - [MainThread] UM.Logger.logException [110]: UM.Settings.ContainerFormatError.ContainerFormatError: When trying to deserialize Flashforge Dreamer, we received an unknown container ID (dreamer)
2021-04-06 23:14:09,675 - ERROR - [MainThread] UM.Settings.ContainerStack.deserialize [429]: When trying to deserialize Flashforge Dreamer, we received an unknown container ID (dreamer)
2021-04-06 23:14:09,678 - ERROR - [MainThread] UM.Logger.logException [106]: Exception: Error in the format of container Flashforge Dreamer: When trying to deserialize Flashforge Dreamer, we received an unknown container ID (dreamer)
2021-04-06 23:14:09,681 - ERROR - [MainThread] UM.Logger.logException [110]: Traceback (most recent call last):
2021-04-06 23:14:09,684 - ERROR - [MainThread] UM.Logger.logException [110]:   File "X:\4.8-exe\build\inst\lib\python3.5\site-packages\UM\Settings\ContainerRegistry.py", line 203, in findContainers
2021-04-06 23:14:09,687 - ERROR - [MainThread] UM.Logger.logException [110]:   File "C:\Program Files\Ultimaker Cura 4.8.0\plugins\LocalContainerProvider\LocalContainerProvider.py", line 75, in loadContainer
2021-04-06 23:14:09,690 - ERROR - [MainThread] UM.Logger.logException [110]:     container.deserialize(f.read(), file_path)
2021-04-06 23:14:09,693 - ERROR - [MainThread] UM.Logger.logException [110]:   File "X:\4.8-exe\build\inst\lib\python3.5\site-packages\cura\Settings\CuraContainerStack.py", line 309, in deserialize
2021-04-06 23:14:09,695 - ERROR - [MainThread] UM.Logger.logException [110]:   File "X:\4.8-exe\build\inst\lib\python3.5\site-packages\UM\Settings\ContainerStack.py", line 430, in deserialize
2021-04-06 23:14:09,698 - ERROR - [MainThread] UM.Logger.logException [110]: UM.Settings.ContainerFormatError.ContainerFormatError: When trying to deserialize Flashforge Dreamer, we received an unknown container ID (dreamer)
2021-04-06 23:14:09,708 - ERROR - [MainThread] UM.Settings.ContainerStack.deserialize [429]: When trying to deserialize Flashforge Dreamer, we received an unknown container ID (dreamer)
2021-04-06 23:14:09,711 - ERROR - [MainThread] UM.Logger.logException [106]: Exception: Error in the format of container Flashforge Dreamer: When trying to deserialize Flashforge Dreamer, we received an unknown container ID (dreamer)
2021-04-06 23:14:09,714 - ERROR - [MainThread] UM.Logger.logException [110]: Traceback (most recent call last):
2021-04-06 23:14:09,717 - ERROR - [MainThread] UM.Logger.logException [110]:   File "X:\4.8-exe\build\inst\lib\python3.5\site-packages\UM\Settings\ContainerRegistry.py", line 203, in findContainers
2021-04-06 23:14:09,720 - ERROR - [MainThread] UM.Logger.logException [110]:   File "C:\Program Files\Ultimaker Cura 4.8.0\plugins\LocalContainerProvider\LocalContainerProvider.py", line 75, in loadContainer
2021-04-06 23:14:09,723 - ERROR - [MainThread] UM.Logger.logException [110]:     container.deserialize(f.read(), file_path)
2021-04-06 23:14:09,726 - ERROR - [MainThread] UM.Logger.logException [110]:   File "X:\4.8-exe\build\inst\lib\python3.5\site-packages\cura\Settings\CuraContainerStack.py", line 309, in deserialize
2021-04-06 23:14:09,728 - ERROR - [MainThread] UM.Logger.logException [110]:   File "X:\4.8-exe\build\inst\lib\python3.5\site-packages\UM\Settings\ContainerStack.py", line 430, in deserialize
2021-04-06 23:14:09,731 - ERROR - [MainThread] UM.Logger.logException [110]: UM.Settings.ContainerFormatError.ContainerFormatError: When trying to deserialize Flashforge Dreamer, we received an unknown container ID (dreamer)
2021-04-06 23:14:09,830 - DEBUG - [Thread-2] cura.Machines.ContainerTree.run [185]: All MachineNode loading completed
2021-04-06 23:14:09,839 - DEBUG - [MainThread] cura.Machines.Models.QualityProfilesDropDownMenuModel._update [62]: Updating QualityProfilesDropDownMenuModel.
2021-04-06 23:14:09,841 - DEBUG - [MainThread] cura.Machines.Models.QualityProfilesDropDownMenuModel._update [74]: No active GlobalStack, set quality profile model as empty.
2021-04-06 23:14:09,844 - DEBUG - [MainThread] cura.Machines.Models.CustomQualityProfilesDropDownMenuModel._update [33]: Updating CustomQualityProfilesDropDownMenuModel.
2021-04-06 23:14:09,846 - DEBUG - [MainThread] cura.Machines.Models.CustomQualityProfilesDropDownMenuModel._update [38]: No active GlobalStack, set CustomQualityProfilesDropDownMenuModel as empty.
2021-04-06 23:14:19,822 - DEBUG - [MainThread] cura.AutoSave._onTimeout [59]: Autosaving preferences, instances and profiles
2021-04-06 23:14:19,839 - ERROR - [MainThread] UM.Settings.ContainerStack.deserialize [429]: When trying to deserialize Flashforge Dreamer, we received an unknown container ID (dreamer)
2021-04-06 23:14:19,841 - ERROR - [MainThread] UM.Logger.logException [106]: Exception: Error in the format of container Flashforge Dreamer: When trying to deserialize Flashforge Dreamer, we received an unknown container ID (dreamer)
2021-04-06 23:14:19,844 - ERROR - [MainThread] UM.Logger.logException [110]: Traceback (most recent call last):
2021-04-06 23:14:19,846 - ERROR - [MainThread] UM.Logger.logException [110]:   File "X:\4.8-exe\build\inst\lib\python3.5\site-packages\UM\Settings\ContainerRegistry.py", line 203, in findContainers
2021-04-06 23:14:19,848 - ERROR - [MainThread] UM.Logger.logException [110]:   File "C:\Program Files\Ultimaker Cura 4.8.0\plugins\LocalContainerProvider\LocalContainerProvider.py", line 75, in loadContainer
2021-04-06 23:14:19,850 - ERROR - [MainThread] UM.Logger.logException [110]:     container.deserialize(f.read(), file_path)
2021-04-06 23:14:19,852 - ERROR - [MainThread] UM.Logger.logException [110]:   File "X:\4.8-exe\build\inst\lib\python3.5\site-packages\cura\Settings\CuraContainerStack.py", line 309, in deserialize
2021-04-06 23:14:19,854 - ERROR - [MainThread] UM.Logger.logException [110]:   File "X:\4.8-exe\build\inst\lib\python3.5\site-packages\UM\Settings\ContainerStack.py", line 430, in deserialize
2021-04-06 23:14:19,856 - ERROR - [MainThread] UM.Logger.logException [110]: UM.Settings.ContainerFormatError.ContainerFormatError: When trying to deserialize Flashforge Dreamer, we received an unknown container ID (dreamer)
2021-04-06 23:14:38,647 - ERROR - [MainThread] UM.Settings.ContainerStack.deserialize [429]: When trying to deserialize Flashforge Dreamer, we received an unknown container ID (dreamer)
2021-04-06 23:14:38,650 - ERROR - [MainThread] UM.Logger.logException [106]: Exception: Error in the format of container Flashforge Dreamer: When trying to deserialize Flashforge Dreamer, we received an unknown container ID (dreamer)
2021-04-06 23:14:38,652 - ERROR - [MainThread] UM.Logger.logException [110]: Traceback (most recent call last):
2021-04-06 23:14:38,654 - ERROR - [MainThread] UM.Logger.logException [110]:   File "X:\4.8-exe\build\inst\lib\python3.5\site-packages\UM\Settings\ContainerRegistry.py", line 203, in findContainers
2021-04-06 23:14:38,657 - ERROR - [MainThread] UM.Logger.logException [110]:   File "C:\Program Files\Ultimaker Cura 4.8.0\plugins\LocalContainerProvider\LocalContainerProvider.py", line 75, in loadContainer
2021-04-06 23:14:38,659 - ERROR - [MainThread] UM.Logger.logException [110]:     container.deserialize(f.read(), file_path)
2021-04-06 23:14:38,661 - ERROR - [MainThread] UM.Logger.logException [110]:   File "X:\4.8-exe\build\inst\lib\python3.5\site-packages\cura\Settings\CuraContainerStack.py", line 309, in deserialize
2021-04-06 23:14:38,663 - ERROR - [MainThread] UM.Logger.logException [110]:   File "X:\4.8-exe\build\inst\lib\python3.5\site-packages\UM\Settings\ContainerStack.py", line 430, in deserialize
2021-04-06 23:14:38,665 - ERROR - [MainThread] UM.Logger.logException [110]: UM.Settings.ContainerFormatError.ContainerFormatError: When trying to deserialize Flashforge Dreamer, we received an unknown container ID (dreamer)
2021-04-06 23:15:08,713 - ERROR - [MainThread] UM.Settings.ContainerStack.deserialize [429]: When trying to deserialize Flashforge Dreamer, we received an unknown container ID (dreamer)
2021-04-06 23:15:08,716 - ERROR - [MainThread] UM.Logger.logException [106]: Exception: Error in the format of container Flashforge Dreamer: When trying to deserialize Flashforge Dreamer, we received an unknown container ID (dreamer)
2021-04-06 23:15:08,719 - ERROR - [MainThread] UM.Logger.logException [110]: Traceback (most recent call last):
2021-04-06 23:15:08,721 - ERROR - [MainThread] UM.Logger.logException [110]:   File "X:\4.8-exe\build\inst\lib\python3.5\site-packages\UM\Settings\ContainerRegistry.py", line 203, in findContainers
2021-04-06 23:15:08,723 - ERROR - [MainThread] UM.Logger.logException [110]:   File "C:\Program Files\Ultimaker Cura 4.8.0\plugins\LocalContainerProvider\LocalContainerProvider.py", line 75, in loadContainer
2021-04-06 23:15:08,725 - ERROR - [MainThread] UM.Logger.logException [110]:     container.deserialize(f.read(), file_path)
2021-04-06 23:15:08,727 - ERROR - [MainThread] UM.Logger.logException [110]:   File "X:\4.8-exe\build\inst\lib\python3.5\site-packages\cura\Settings\CuraContainerStack.py", line 309, in deserialize
2021-04-06 23:15:08,729 - ERROR - [MainThread] UM.Logger.logException [110]:   File "X:\4.8-exe\build\inst\lib\python3.5\site-packages\UM\Settings\ContainerStack.py", line 430, in deserialize
2021-04-06 23:15:08,731 - ERROR - [MainThread] UM.Logger.logException [110]: UM.Settings.ContainerFormatError.ContainerFormatError: When trying to deserialize Flashforge Dreamer, we received an unknown container ID (dreamer)
2021-04-06 23:15:38,773 - ERROR - [MainThread] UM.Settings.ContainerStack.deserialize [429]: When trying to deserialize Flashforge Dreamer, we received an unknown container ID (dreamer)
2021-04-06 23:15:38,775 - ERROR - [MainThread] UM.Logger.logException [106]: Exception: Error in the format of container Flashforge Dreamer: When trying to deserialize Flashforge Dreamer, we received an unknown container ID (dreamer)
2021-04-06 23:15:38,778 - ERROR - [MainThread] UM.Logger.logException [110]: Traceback (most recent call last):
2021-04-06 23:15:38,780 - ERROR - [MainThread] UM.Logger.logException [110]:   File "X:\4.8-exe\build\inst\lib\python3.5\site-packages\UM\Settings\ContainerRegistry.py", line 203, in findContainers
2021-04-06 23:15:38,782 - ERROR - [MainThread] UM.Logger.logException [110]:   File "C:\Program Files\Ultimaker Cura 4.8.0\plugins\LocalContainerProvider\LocalContainerProvider.py", line 75, in loadContainer
2021-04-06 23:15:38,784 - ERROR - [MainThread] UM.Logger.logException [110]:     container.deserialize(f.read(), file_path)
2021-04-06 23:15:38,786 - ERROR - [MainThread] UM.Logger.logException [110]:   File "X:\4.8-exe\build\inst\lib\python3.5\site-packages\cura\Settings\CuraContainerStack.py", line 309, in deserialize
2021-04-06 23:15:38,788 - ERROR - [MainThread] UM.Logger.logException [110]:   File "X:\4.8-exe\build\inst\lib\python3.5\site-packages\UM\Settings\ContainerStack.py", line 430, in deserialize
2021-04-06 23:15:38,790 - ERROR - [MainThread] UM.Logger.logException [110]: UM.Settings.ContainerFormatError.ContainerFormatError: When trying to deserialize Flashforge Dreamer, we received an unknown container ID (dreamer)

**Comment by HadiSDev** [comment]
Nvm I fixed it. Missing base json

**Comment by KeltE** [comment]
I know base.def.json is missing. profile is made on top of Dreamer NX profile and uses "flashforge_base.def.json" file and qualities. Since now Cura 4.9 Beta is also out, with the addition of Flashforge dreamer NX profile already.

![image](https://user-images.githubusercontent.com/66703581/113781288-22efc100-9739-11eb-811b-a7df6ca76a24.png)

**Comment by HadiSDev** [comment]
I tried to print but the printer says "File content error". This is the gx file
[D_calicat.zip](https://github.com/KeltE/Flashforge_Dreamer_Cura-definition_concept/files/6279427/D_calicat.zip)

**Comment by KeltE** [comment]
Thank you very much for your feedback! I tested this Gcode. I got to work when I removed the T0 command made at the beginning of Cura. And I also discovered that the retraction is for some reason 6.5 mm. Although I think I changed it to 1.3 mm.

> I tried to print but the printer says "File content error". This is the gx file
> [D_calicat.zip](https://github.com/KeltE/Flashforge_Dreamer_Cura-definition_concept/files/6279427/D_calicat.zip)

**Comment by HadiSDev** [comment]
How do we remove the T0 command

**Comment by KeltE** [comment]
Hard to say. As it turns out, this has also been a problem with 2 nozzle printers before. Others have had this problem and I have seen the topic here on Github. The best solution would be to integrate it into a .GX plugin. In the form of a script that removes it. But my knowledge ends here. :)

**Comment by KeltE** [comment]
I opened a new issue where I got this GXplugin.
